libkbfs/leveled_mutex.go

```diff
@@ -24,27 +25,23 @@ import (
 // accessed exclusively. It immediately panics upon any lock
 // contention.
 type exclusiveLock struct {
-	// TODO: This can be replaced with an atomic value and
-	// compare-and-swap.
-	ch chan struct{}
+	v *int32
 }
 
 func makeExclusiveLock() exclusiveLock {
 	return exclusiveLock{
-		ch: make(chan struct{}, 1),
+		v: new(int32),
 	}
 }
 
 func (l exclusiveLock) lock() {
-	select {
-	case l.ch <- struct{}{}:
-	default:
+	if !atomic.CompareAndSwapInt32(l.v, 0, 1) {
 		panic("unexpected concurrent access")
 	}
 }
 
 func (l exclusiveLock) unlock() {
-	<-l.ch
+	atomic.StoreInt32(l.v, 0)
```

Comment:
can you make this a CAS operation too? (from 1 to 0) That way we'd catch double-unlocks. (The old version should have done something similar, too.)